src/Disks/IDiskRemote.h

```diff
@@ -0,0 +1,250 @@
+#pragma once
+#include <Common/config.h>
+
+#if USE_AWS_S3
+
+#include <atomic>
+#include "Disks/DiskFactory.h"
+#include "Disks/Executor.h"
+#include <Poco/DirectoryIterator.h>
+#include <utility>
+#include <aws/s3/model/CopyObjectRequest.h>
+#include <aws/s3/model/DeleteObjectsRequest.h>
+#include <aws/s3/model/GetObjectRequest.h>
+#include <aws/s3/model/ListObjectsV2Request.h>
+#include <aws/s3/model/HeadObjectRequest.h>
+#include <Common/MultiVersion.h>
+
+
+namespace DB
+{
+
+/// Helper class to collect keys into chunks of maximum size (to prepare batch requests to AWS API).
+/// Used for both S3 and HDFS.
+class RemoteFSPathKeeper : public std::list<Aws::Vector<Aws::S3::Model::ObjectIdentifier>>
+{
+public:
+    void addPath(const String & path)
+    {
+        if (empty() || back().size() >= chunk_limit)
+        { /// add one more chunk
+            push_back(value_type());
+            back().reserve(chunk_limit);
+        }
+
+        Aws::S3::Model::ObjectIdentifier obj;
+        obj.SetKey(path);
+        back().push_back(obj);
+    }
+
+private:
+    /// limit for one DeleteObject request
+    /// see https://docs.aws.amazon.com/AmazonS3/latest/API/API_DeleteObjects.html
+    const static size_t chunk_limit = 1000;
```

Comment:
chunk_limit size for `RemoteFSPathKeeper` should be configurable and comment regarding it should be a part of DiskS3 class.